### PR TITLE
fix!: drop nil votes in misbehaviour handling

### PR DESCRIPTION
**Pull request by sainoe**
## Description

Closes: #1403

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [ ] Included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] Added `!` to the type prefix if state-machine breaking change (i.e., requires coordinated upgrade)
* [ ] Confirmed this PR does not introduce changes requiring state migrations, OR migration code has been added to consumer and/or provider modules
* [ ] Targeted the correct branch (see [PR Targeting](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#pr-targeting))
* [ ] Provided a link to the relevant issue or specification
* [ ] Followed the guidelines for [building SDK modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules)
* [ ] Included the necessary unit and integration [tests](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#testing)
* [ ] Added a changelog entry to `CHANGELOG.md`
* [ ] Included comments for [documenting Go code](https://blog.golang.org/godoc)
* [ ] Updated the relevant documentation or specification
* [ ] Reviewed "Files changed" and left comments if necessary 
* [ ] Confirmed all CI checks have passed
* [ ] If this PR is library API breaking, bump the go.mod version string of the repo, and follow through on a new major release for both the consumer and provider

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

* [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed `!` in the type prefix if API or client breaking change
* [ ] confirmed this PR does not introduce changes requiring state migrations, OR confirmed migration code has been added to consumer and/or provider modules
* [ ] confirmed all author checklist items have been addressed
* [ ] reviewed state machine logic
* [ ] reviewed API design and naming
* [ ] reviewed documentation is accurate
* [ ] reviewed tests and test coverage

### CODE REVIEW

File: tests/integration/misbehaviour.go [hunk 1]
```diff
@@ -223,21 +261,17 @@ func (s *CCVTestSuite) TestGetByzantineValidators() {
 				s.NoError(err)
 				s.Equal(len(tc.expByzantineValidators), len(byzantineValidators))
 
-				// For both lunatic and equivocation attacks all the validators
-				// who signed the bad header (Header2) should be in returned in the evidence
+				// For both lunatic and equivocation attacks, all the validators
+				// who signed the bad header and didn't vote nil  should be returned
```
Comment (insumity) [inline tests/integration/misbehaviour.go:265]:
```suggestion
				// who signed both headers and didn't vote nil should be returned
```

File: testutil/crypto/evidence.go [hunk 2]
```diff
@@ -89,3 +91,37 @@ func MakeAndSignVoteWithForgedValAddress(
 	vote.Signature = v.Signature
 	return vote
 }
+
+// UpdateHeaderCommitWithNilVotes updates the given light client header
+// by changing the commit BlockIDFlag of the given validators to nil
+//
+// Note that this method is solely used for testing purpose
```
Comment (insumity) [inline testutil/crypto/evidence.go:98]:
```suggestion
// Note that this method is solely used for testing purposes
```

File: testutil/crypto/evidence.go [hunk 3]
```diff
@@ -89,3 +91,37 @@ func MakeAndSignVoteWithForgedValAddress(
...
+	}
+
+	valset, err := tmtypes.ValidatorSetFromProto(header.ValidatorSet)
+	if err != nil {
+		panic(err)
+	}
+
+	for _, v := range validators {
+		// get validator index in valset
+		idx, _ := valset.GetByAddress(v.Address)
```
Comment (insumity) [inline testutil/crypto/evidence.go:117]: does it make sense to check if `idx == -1`?

Comment (sainoe) [reply testutil/crypto/evidence.go:117]: We assume that all passed validators are part of the valset, but true we can add a condition to handle this case.

Comment (sainoe) [reply testutil/crypto/evidence.go:117]: Changed in `[52a9421d88f9](https://github.com/cosmos/interchain-security/pull/1404/commits/52a9421d88f9cca3c8463aa28bfc3086c9a68213)`

**Comment by insumity** [comment]
Looks good to me in regards to the `nil` vote changes. Thanks!

Just some questions:

1. From what I understand, this PR does not introduce a test with the case where we have a `nil` vote with an incorrect signature from a validator `v` that tests that now `v` is not getting slashed, right? Do we want to add such a test?

2. Do we intend to modify the misbehaviour check to only consider ICS light clients as well?

**Comment by sainoe** [comment]
> From what I understand, this PR does not introduce a test with the case where we have a nil vote with an incorrect signature from a validator v that tests that now v is not getting slashed, right? Do we want to add such a test?

I can't change the validator's commit sig to have a nil vote and an incorrect signature as well, but the cometBFT behaviour will be the same since nil vote aren't verified.

> Do we intend to modify the misbehaviour check to only consider ICS light clients as well?

Good point let me add this quickly.

File: x/ccv/provider/keeper/misbehaviour.go [hunk 4]
```diff
@@ -142,8 +142,20 @@ func headerToLightBlock(h ibctmtypes.Header) (*tmtypes.LightBlock, error) {
 }
 
 // CheckMisbehaviour checks that headers in the given misbehaviour forms
-// a valid light client attack and that the corresponding light client isn't expired
+// a valid light client attack from an ICS consumer chain and that the light client isn't expired
 func (k Keeper) CheckMisbehaviour(ctx sdk.Context, misbehaviour ibctmtypes.Misbehaviour) error {
+	// check that the misbehaviour is for an ICS consumer chain
+	clientID, found := k.GetConsumerClientId(ctx, misbehaviour.Header1.Header.ChainID)
```
Comment (insumity) [inline x/ccv/provider/keeper/misbehaviour.go:148]: nit: Maybe rename `clientID` to `clientId` to be consistent with the `GetConsumerClientId`.

File: x/ccv/provider/keeper/misbehaviour.go [hunk 5]
```diff
@@ -142,8 +142,20 @@ func headerToLightBlock(h ibctmtypes.Header) (*tmtypes.LightBlock, error) {
...
+	if !found {
+		return fmt.Errorf("incorrect misbehaviour with conflicting headers from a non-existent consumer chain: %s", misbehaviour.Header1.Header.ChainID)
+	} else if misbehaviour.ClientId != clientID {
```
Comment (insumity) [inline x/ccv/provider/keeper/misbehaviour.go:151]: Nice! Maybe we can be consistent on the way we retrieve the `ClientId` from `misbehaviour`. Here we do `.ClientId` but later we do `GetClientID()`.

File: x/ccv/provider/keeper/misbehaviour.go [hunk 6]
```diff
@@ -142,8 +142,20 @@ func headerToLightBlock(h ibctmtypes.Header) (*tmtypes.LightBlock, error) {
...
+		return fmt.Errorf("incorrect misbehaviour: expected client ID for consumer chain %s is %s got %s",
+			misbehaviour.Header1.Header.ChainID,
+			clientID,
+			misbehaviour.ClientId,
+		)
+	}
+
 	clientState, found := k.clientKeeper.GetClientState(ctx, misbehaviour.GetClientID())
```
Comment (insumity) [inline x/ccv/provider/keeper/misbehaviour.go:159]: nit: We could use directly the `clientId` here. 
```suggestion
	clientState, found := k.clientKeeper.GetClientState(ctx, clientId)
```
This way we would avoid a potential regression where the previous `else if misbehaviour.ClientId != clientID` statement is accidentally removed and `misbehaviour.GetClientID()` is something else.

Comment (insumity) [reply x/ccv/provider/keeper/misbehaviour.go:159]: Or it might make sense to add some sanity tests for this to prevent a possible regression.

**Review by insumity — APPROVED** [review]
Left some minor comments, but LGTM for merging.

File: x/ccv/provider/keeper/misbehaviour.go (hunk 4 above)
Comment (sainoe) [reply x/ccv/provider/keeper/misbehaviour.go]: Sure changed in last commit

File: x/ccv/provider/keeper/misbehaviour.go [hunk 7]
```diff
@@ -142,9 +142,21 @@ func headerToLightBlock(h ibctmtypes.Header) (*tmtypes.LightBlock, error) {
...
+		return fmt.Errorf("incorrect misbehaviour with conflicting headers from a non-existent consumer chain: %s", misbehaviour.Header1.Header.ChainID)
+	} else if misbehaviour.ClientId != clientId {
+		return fmt.Errorf("incorrect misbehaviour: expected client ID for consumer chain %s is %s got %s",
+			misbehaviour.Header1.Header.ChainID,
+			clientId,
+			misbehaviour.ClientId,
+		)
+	}
+
+	clientState, found := k.clientKeeper.GetClientState(ctx, clientId)
```
Comment (insumity) [inline x/ccv/provider/keeper/misbehaviour.go:159]: Below in:
```
	clientStore := k.clientKeeper.ClientStore(ctx, misbehaviour.GetClientID())
```
there's discrepancy in how we use `misbehaviour.GetClientID()` versus `misbehaviour.ClientId` versus simply `clientId`. I believe it makes sense to use everywhere `clientId`.

**Pull request merged**